### ISSUE TITLE
Any plans on integrating ArUco tracking?

### ISSUE THREAD

**Issue by 9and3** [maluoi/StereoKit#115]
First things first: what an amazing tool StereoKit is, thanks!
I was also wondering if there were any plans on integrating ArUco on StereoKit. I would need those to anchor/position my holograms in correspondence with fiducial markers.
Also if anyone has an idea on how to implement this on StereoKit I'm here :)

**Comment by maluoi** [comment]
So I suppose the first thing to highlight is that if you can use Microsoft's [QR code tracking library](https://docs.microsoft.com/en-us/windows/mixed-reality/develop/platform-capabilities-and-apis/qr-code-tracking), that's [super easy to hook into](https://stereokit.net/Pages/Guides/Using-QR-Codes.html)!

I am planning on adding a proper anchoring system to StereoKit at some point in the near future. There's a lot of different tech and strategies out there for anchoring holograms, and so I need a good way to surface up the entire concept! The biggest question on my mind though, is where should each anchor type exist in the code-base? For OpenXR provided anchors, it's super appropriate for them to be core features that ship with StereoKit! For things like the Microsoft QR code library, it makes more sense as a plugin that provides information to StereoKit's future anchoring system as an external source.

For ArUco markers, I'm not so sure I'd want them as a core feature. The dependencies would add a fair bit of binary weight, and add some additional build complexity that I don't want. StereoKit is also a more future thinking library, and I think it's safe to say most consumer-ready anchoring systems will be markerless! But ArUco's are still pretty practical, so I may indeed want to make sure there's a plugin for it at the very least :)

I'd also love to know a little more about how you'd intend to use the ArUco markers! Are you interested in like... real-time object tracking, metadata for your environment, something else? What sort of hardware would you want to track it with, would this be like a Vive Pro camera?

Anyhow, sorry for the long reply, questions like these are great for helping me to settle things in my mind too :)

**Comment by 9and3** [comment]
Hello Nick, thank you for your very interesting answer, it's my pleasure to read your comment. 

Concerning the QR code tracking library is where all my frustrations as a not-so-experienced developer in Hololens and Unity come from. There is a lot of QR trackers out there but no one of them seems to work: maybe cause of project settings, maybe of an MRTK profile, maybe of a wrong version or sub-version of Unity, who knows? This is one of the reasons why I would like to start building all my projects on StereoKit :) 

I do understand completely your stance on keeping StereoKit light by avoiding any external dependency, this is indeed one of the fascinating features of StereoKit. An ArUco plug-in would be amazing though!

To answer your question about why I need those.. I am actually interested in real-time object tracking. I am a researcher working on augmented fabrication and I was thinking of ArUco to track both my piece to process (sometimes you have to turn your timber element or move it, etc'..) and my tools also. Since I would need fiducial markers on my tools, they should be fairly small, that's why I was thinking about ArUco and not QR code. 

In relation to the anchoring, for example in my case, are absolutely fundamental because the operator would need to have a maximal level of precision in terms of holographic stability. That's why I was asking for the anchors, this is, to be frank, the missing piece I am waiting to burn everything and get a new fresh start in StereoKit. 
(please tell me if I am missing something here about holographic stability in StereoKit)

**Comment by maluoi** [comment]
Ok, thanks for the details, that's super helpful! Your example is definitely a good illustration of how ArUco markers could be practical for a lot longer than I was initially thinking, so this definitely bumped the importance in my mind.

And you're totally right, it's tough to find a good tracking library and get it working well, Unity often makes this even trickier by breaking in strange ways while doing it! I actually did some of this myself too, I worked on Spectator View a couple years ago, which used an ArUco marker to spatially sync a HoloLens to a phone running AR Core/Kit. That was... brutal.

FYI on timeline! There's nothing currently stopping an ArUco library from working with StereoKit right now, if you can find one that behaves. If you're using StereoKit from the C++ side, that might even simplify builds, or let you hook into a few other libraries! I'm currently tracking the planned anchoring system as #78, and I probably won't be looking into writing any tracking plugins myself until that's finished (maybe in a month or two?).

**Comment by abdullahthabit** [comment]
Hi @maluoi,

First of all, thank you for StereoKit! 
To follow up with @9and3  question, I am also interested in using Aruco with StereoKit for positioning Holograms and tracking tools in surgical settings. But I don't know how to go about accessing the camera streams to get the frames for processing. Any suggestions on how to do this with StereoKit/OpenXR? I am interested in using the Hololens2 with mainly the PV camera for tracking. 

Cheers,

**Comment by maluoi** [comment]
Through UWP, camera streams are accessible through the MediaCapture class in the Windows.Media.Capture namespace! SpectatorView is a bit of a complex project, but they [work with the MediaCapture class](https://github.com/microsoft/MixedReality-SpectatorView/blob/master/src/SpectatorView.Unity/Assets/PhotoCapture/Scripts/HoloLensCamera.cs) as well as [Aruco markers](https://github.com/microsoft/MixedReality-SpectatorView/tree/7796da6acb0ae41bed1b9e0e9d1c5c683b4b8374/src/SpectatorView.Native). I haven't worked with this code-base since it got re-written, but it could be a decent reference if you're trying to go through this yourself!

Aruco is still done on the native side, so there is still the issue with working across the language barrier. It's complicated stuff, StereoKit's affinity with Visual Studio may make pieces of this a little easier, but I don't expect it'll be terribly fun.

**Comment by abdullahthabit** [comment]
Thank you Nick for the suggestions! Yes I have come across the MediaCapture class earlier but thought maybe there's an easier way within the OpenXR API to access the camera streams, so that I don't go through any conflict issues, especially that I am not very experienced with this. Would you mind if I came back to you with other questions, and would it be better to reach out to you here or via discord or other channels?

Cheers,

**Comment by maluoi** [comment]
Yeah, I think this runs into some interesting issues about what OpenXR should and shouldn't be responsible for. In the case of camera access, I imagine they'd say it's outside of their realm of responsibility. I think they'd be more likely to add an Aruco tracking extension than a camera streaming extension, if that makes sense.

C# does have a nice ecosystem though, so I'm sure there are other things out there that could make life easier too? Like here's an interesting [OpenCV for C#](https://github.com/shimat/opencvsharp) library! And for general webcam access, they literally have top 20 lists for nuget packages! I only recommend MediaCapture because I know for sure it can get you the right data :)

And yeah, feel free to ask away! I'm happy to answer questions wherever you're most comfortable. Github Issues/Discussions is a great permanent record, and you'll often get a lot longer form answers from me. Discord goes a little quicker, but is very short-form. I'll see it either way :)
